File: src/main/go/aoc2019/day03.go
package main

import (
	"aoc2019/utl"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Run03 Day 3: Crossed Wires
func Run03() {

	t := time.Now()
	day03Part1()

	fmt.Printf("=== end (%d ms) ===\n", time.Now().Sub(t).Milliseconds())
}

func wirePoints(w []string) []Point {
	r := make([]Point, 0, 1000)
	p := Point{0, 0}
	for _, wi := range w {
		l, _ := strconv.Atoi(wi[1:])
		for i := 0; i < l; i++ {
			switch wi[0] {
			case 'U':
				p = p.north()
			case 'D':
				p = p.south()
			case 'L':
				p = p.west()
			case 'R':
				p = p.east()
			}
			r = append(r, p)
		}
	}
	return r
}

func day03Part1() {
	s := utl.ReadInputStrings("input03.txt")
	w1 := wirePoints(strings.Split(s[0], ","))
	w2 := wirePoints(strings.Split(s[1], ","))

	fmt.Println(len(w1))
	fmt.Println(len(w2))
	inters := intersect(w1, w2)
	// fmt.Println(inters)

	p := inters[0]
	for _, x := range inters {
		if x.manhattan() < p.manhattan() {
			p = x
		}
	}
	fmt.Println(p.manhattan())

	wlen := func(p Point) int {
		return p.indexIn(w1) + p.indexIn(w2) + 2
	}

	fmt.Println("--- part 2 ---")
	p = inters[0]
	for _, x := range inters {
		if wlen(x) < wlen(p) {
			p = x
		}
	}
	fmt.Println(wlen(p))
}

func intersect(w1, w2 []Point) []Point {
	m := make(map[Point]bool, len(w1))
	for _, p := range w1 {
		m[p] = true
	}
	result := make([]Point, 0, len(w2))
	seen := make(map[Point]bool) // avoid dups
	for _, p := range w2 {
		if m[p] && (!seen[p]) {
			result = append(result, p)
			seen[p] = true
		}
	}
	return result
}

File: src/main/csharp/aoc2019/Day03.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using aoc;

namespace aoc2019
{
    class Day03 : AocDay
    {

        string[] Input1 = ReadLines("input03.txt")[0].Split(',');
        string[] Input2 = ReadLines("input03.txt")[1].Split(',');

        public void main()
        {
            Console.WriteLine("=== AOC 2019 day 03");
            var t0 = DateTime.Now;

            Console.WriteLine("--- part 1");
            Part1();

            Console.WriteLine("--- part 2");
            Part2();

            Console.WriteLine("=== end ({0} ms)", (DateTime.Now - t0).Milliseconds);
        }

        private void Part1()
        {
            var l1 = WirePoints(Input1).ToArray();
            var l2 = WirePoints(Input2).ToArray();
            var crossing = l1.ToHashSet().Intersect(l2.ToHashSet()).Min(p => p.Manhattan());
            Console.WriteLine(crossing);
        }

        private void Part2()
        {
            var l1 = WirePoints(Input1).ToArray();
            var l2 = WirePoints(Input2).ToArray();
            var crossings = l1.ToHashSet().Intersect(l2.ToHashSet());
            var p = crossings.Select(p => 2 + Array.IndexOf(l1, p) + Array.IndexOf(l2, p)).Min();
            Console.WriteLine(p);
        }

        private IEnumerable<Point> WirePoints(string[] input)
        {
            var p = Point.Of(0, 0);
            foreach (string s in input)
            {
                var dir = s.ElementAt(0);
                var len = int.Parse(s.Substring(1));
                while (len > 0)
                {
                    switch (dir)
                    {
                        case 'U': yield return (p = Point.Of(p.X, p.Y - 1)); break;
                        case 'D': yield return (p = Point.Of(p.X, p.Y + 1)); break;
                        case 'L': yield return (p = Point.Of(p.X - 1, p.Y)); break;
                        case 'R': yield return (p = Point.Of(p.X + 1, p.Y)); break;

                        default: throw new InvalidDataException($"dir {dir} unknown");
                    }
                    len--;
                }

            }
        }
    }
}
